### ISSUE TITLE
suboptimal memory usage of repetition operator

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13793]
Migrated from [rt.perl.org#121780](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=121780) (status was 'open')

Searchable as RT121780$

**Comment by p5pRT** [comment]
### From efimov@reg.ru

suboptimal memory usage of repetition operator

===
my $s1;
$s1 \.= "x" for \(1\.\.200\*1024\*1024\);
print "ok\\n";
sleep 1000;
$s1 \.= '';
===
Uses 200Mb \(during sleep\(\)\)


===
my $s1;
$s1 = "x" x \(200\*1024\*1024\);
print "ok\\n";
sleep 1000;
$s1 \.= '';
===
Uses 400Mb \(during sleep\(\)\)

reproduced in 5\.14 and 5\.19

perl \-V
Summary of my perl5 \(revision 5 version 14 subversion 2\) configuration&#8203;:

&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.6\.42\-37\-generic\,
archname=x86\_64\-linux\-gnu\-thread\-multi
&nbsp;    uname='linux panlong 2\.6\.42\-37\-generic \#58\-ubuntu smp thu jan 24
15&#8203;:28&#8203;:10 utc 2013 x86\_64 x86\_64 x86\_64 gnulinux '
&nbsp;    config\_args='\-Dusethreads \-Duselargefiles \-Dccflags=\-DDEBIAN
\-Dcccdlflags=\-fPIC \-Darchname=x86\_64\-linux\-gnu \-Dprefix=/usr
\-Dprivlib=/usr/share/perl/5\.14 \-Darchlib=/usr/lib/perl/5\.14
\-Dvendorprefix=/usr \-Dvendorlib=/usr/share/perl5
\-Dvendorarch=/usr/lib/perl5 \-Dsiteprefix=/usr/local
\-Dsitelib=/usr/local/share/perl/5\.14\.2
\-Dsitearch=/usr/local/lib/perl/5\.14\.2 \-Dman1dir=/usr/share/man/man1
\-Dman3dir=/usr/share/man/man3 \-Dsiteman1dir=/usr/local/man/man1
\-Dsiteman3dir=/usr/local/man/man3 \-Duse64bitint \-Dman1ext=1
\-Dman3ext=3perl \-Dpager=/usr/bin/sensible\-pager \-Uafs \-Ud\_csh
\-Ud\_ualarm \-Uusesfio \-Uusenm \-Ui\_libutil \-DDEBUGGING=\-g \-Doptimize=\-O2
\-Duseshrplib \-Dlibperl=libperl\.so\.5\.14\.2 \-des'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN
\-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include
\-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2 \-g'\,
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fno\-strict\-aliasing
\-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.6\.3'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\,
lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib
/usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib /usr/lib
&nbsp;    libs=\-lgdbm \-lgdbm\_compat \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    perllibs=\-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    libc=\, so=so\, useshrplib=true\, libperl=libperl\.so\.5\.14\.2
&nbsp;    gnulibc\_version='2\.15'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-O2 \-g \-L/usr/local/lib
\-fstack\-protector'


Characteristics of this binary \(from libperl\)&#8203;:
&nbsp;  Compile\-time options&#8203;: MULTIPLICITY PERL\_DONT\_CREATE\_GVSV
&nbsp;                        PERL\_IMPLICIT\_CONTEXT PERL\_MALLOC\_WRAP
&nbsp;                        PERL\_PRESERVE\_IVUV USE\_64\_BIT\_ALL USE\_64\_BIT\_INT
&nbsp;                        USE\_ITHREADS USE\_LARGE\_FILES USE\_PERLIO USE\_PERL\_ATOF
&nbsp;                        USE\_REENTRANT\_API
&nbsp;  Locally applied patches&#8203;:
&nbsp;    DEBPKG&#8203;:debian/arm\_thread\_stress\_timeout \-
http&#8203;://bugs\.debian\.org/501970 Raise the timeout of
ext/threads/shared/t/stress\.t to accommodate slower build hosts
&nbsp;    DEBPKG&#8203;:debian/cpan\_definstalldirs \- Provide a sensible INSTALLDIRS
default for modules installed from CPAN\.
&nbsp;    DEBPKG&#8203;:debian/db\_file\_ver \- http&#8203;://bugs\.debian\.org/340047 Remove
overly restrictive DB\_File version check\.
&nbsp;    DEBPKG&#8203;:debian/doc\_info \- Replace generic man\(1\) instructions with
Debian\-specific information\.
&nbsp;    DEBPKG&#8203;:debian/enc2xs\_inc \- http&#8203;://bugs\.debian\.org/290336 Tweak
enc2xs to follow symlinks and ignore missing @&#8203;INC directories\.
&nbsp;    DEBPKG&#8203;:debian/libnet\_config\_path \- Set location of libnet\.cfg to
/etc/perl/Net as /usr may not be writable\.
&nbsp;    DEBPKG&#8203;:debian/m68k\_thread\_stress \- http&#8203;://bugs\.debian\.org/517938
http&#8203;://bugs\.debian\.org/495826 Disable some threads tests on m68k for
now due to missing TLS\.
&nbsp;    DEBPKG&#8203;:debian/mod\_paths \- Tweak @&#8203;INC ordering for Debian
&nbsp;    DEBPKG&#8203;:debian/module\_build\_man\_extensions \-
http&#8203;://bugs\.debian\.org/479460 Adjust Module&#8203;::Build manual page
extensions for the Debian Perl policy
&nbsp;    DEBPKG&#8203;:debian/prune\_libs \- http&#8203;://bugs\.debian\.org/128355 Prune the
list of libraries wanted to what we actually need\.
&nbsp;    DEBPKG&#8203;:fixes/net\_smtp\_docs \- \[rt\.cpan\.org \#36038\]
http&#8203;://bugs\.debian\.org/100195 Document the Net&#8203;::SMTP 'Port' option
&nbsp;    DEBPKG&#8203;:debian/perlivp \- http&#8203;://bugs\.debian\.org/510895 Make perlivp
skip include directories in /usr/local
&nbsp;    DEBPKG&#8203;:debian/disable\-zlib\-bundling \- Disable zlib bundling in
Compress&#8203;::Raw&#8203;::Zlib
&nbsp;    DEBPKG&#8203;:debian/cpanplus\_definstalldirs \-
http&#8203;://bugs\.debian\.org/533707 Configure CPANPLUS to use the site
directories by default\.
&nbsp;    DEBPKG&#8203;:debian/cpanplus\_config\_path \- Save local versions of
CPANPLUS&#8203;::Config&#8203;::System into /etc/perl\.
&nbsp;    DEBPKG&#8203;:debian/deprecate\-with\-apt \- http&#8203;://bugs\.debian\.org/580034
Point users to Debian packages of deprecated core modules
&nbsp;    DEBPKG&#8203;:fixes/hurd\-ccflags \- \[a190e64\]
http&#8203;://bugs\.debian\.org/587901 \[perl \#92244\] Make hints/gnu\.sh append
to $ccflags rather than overriding them
&nbsp;    DEBPKG&#8203;:debian/squelch\-locale\-warnings \-
http&#8203;://bugs\.debian\.org/508764 Squelch locale warnings in Debian
package maintainer scripts
&nbsp;    DEBPKG&#8203;:debian/skip\-upstream\-git\-tests \- Skip tests specific to the
upstream Git repository
&nbsp;    DEBPKG&#8203;:fixes/extutils\-cbuilder\-cflags \- \[011e8fb\]
http&#8203;://bugs\.debian\.org/624460 \[perl \#89478\] Append CFLAGS and LDFLAGS
to their Config\.pm counterparts in EU&#8203;::CBuilder
&nbsp;    DEBPKG&#8203;:fixes/module\-build\-home\-directory \-
http&#8203;://bugs\.debian\.org/624850 \[rt\.cpan\.org \#67893\] Fix failing tilde
test when run under a UID without a passwd entry
&nbsp;    DEBPKG&#8203;:debian/patchlevel \- http&#8203;://bugs\.debian\.org/567489 List
packaged patches for 5\.14\.2\-6ubuntu2\.4 in patchlevel\.h
&nbsp;    DEBPKG&#8203;:fixes/h2ph\-multiarch \- \[e7ec705\]
http&#8203;://bugs\.debian\.org/625808 \[perl \#90122\] Make h2ph correctly search
gcc include directories
&nbsp;    DEBPKG&#8203;:fixes/index\-tainting \- \[3b36395\]
http&#8203;://bugs\.debian\.org/291450 \[perl \#64804\] RT 64804&#8203;: tainting with
index\(\) of a constant
&nbsp;    DEBPKG&#8203;:debian/skip\-kfreebsd\-crash \- http&#8203;://bugs\.debian\.org/628493
\[perl \#96272\] Skip a crashing test case in t/op/threads\.t on
GNU/kFreeBSD
&nbsp;    DEBPKG&#8203;:fixes/document\_makemaker\_ccflags \-
http&#8203;://bugs\.debian\.org/628522 \[rt\.cpan\.org \#68613\] Document that
CCFLAGS should include $Config\{ccflags\}
&nbsp;    DEBPKG&#8203;:fixes/sys\-syslog\-socket\-timeout\-kfreebsd\.patch \-
http&#8203;://bugs\.debian\.org/627821 \[rt\.cpan\.org \#69997\] Use a socket
timeout on GNU/kFreeBSD to catch ICMP port unreachable messages
&nbsp;    DEBPKG&#8203;:fixes/hurd\-hints \- http&#8203;://bugs\.debian\.org/636609 Improve
general GNU hints\, needed for GNU/Hurd\.
&nbsp;    DEBPKG&#8203;:fixes/pod\_fixes \- \[7698aed\] http&#8203;://bugs\.debian\.org/637816
Fix typos in several pod/perl\*\.pod files
&nbsp;    DEBPKG&#8203;:debian/find\_html2text \- http&#8203;://bugs\.debian\.org/640479
Configure CPAN&#8203;::Distribution with correct name of html2text
&nbsp;    DEBPKG&#8203;:fixes/digest\_eval\_hole \- http&#8203;://bugs\.debian\.org/644108
Close the eval "require $module" security hole in
Digest\->new\($algorithm\)
&nbsp;    DEBPKG&#8203;:fixes/hurd\-ndbm \- \[f0d0a20\] \[perl \#102680\]
http&#8203;://bugs\.debian\.org/645989 Add GNU/Hurd hints for NDBM\_File
&nbsp;    DEBPKG&#8203;:fixes/sysconf\.t\-posix \- \[8040185\] \[perl \#102888\]
http&#8203;://bugs\.debian\.org/646016 Fix hang in ext/POSIX/t/sysconf\.t on
GNU/Hurd
&nbsp;    DEBPKG&#8203;:fixes/hurd\-largefile \- \[1fda587\] \[perl \#103014\]
http&#8203;://bugs\.debian\.org/645790 enable LFS on GNU/Hurd
&nbsp;    DEBPKG&#8203;:debian/hurd\_test\_todo\_syslog \-
http&#8203;://bugs\.debian\.org/650093 Disable failing GNU/Hurd tests in
cpan/Sys\-Syslog/t/syslog\.t
&nbsp;    DEBPKG&#8203;:fixes/hurd\_skip\_itimer\_virtual \- \[rt\.cpan\.org \#72754\]
http&#8203;://bugs\.debian\.org/650094 Skip interval timer tests in Time&#8203;::HiRes
on GNU/Hurd
&nbsp;    DEBPKG&#8203;:debian/hurd\_test\_skip\_socketpair \-
http&#8203;://bugs\.debian\.org/650186 Disable failing GNU/Hurd tests
ext/Socket/t/socketpair\.t
&nbsp;    DEBPKG&#8203;:debian/hurd\_test\_skip\_sigdispatch \-
http&#8203;://bugs\.debian\.org/650188 Disable failing GNU/Hurd tests
op/sigdispatch\.t
&nbsp;    DEBPKG&#8203;:debian/hurd\_test\_skip\_stack \- http&#8203;://bugs\.debian\.org/650175
Disable failing GNU/Hurd tests dist/threads/t/stack\.t
&nbsp;    DEBPKG&#8203;:debian/hurd\_test\_skip\_recv \- http&#8203;://bugs\.debian\.org/650095
Disable failing GNU/Hurd tests cpan/autodie/t/recv\.t
&nbsp;    DEBPKG&#8203;:debian/hurd\_test\_skip\_libc \- http&#8203;://bugs\.debian\.org/650097
Disable failing GNU/Hurd tests dist/threads/t/libc\.t
&nbsp;    DEBPKG&#8203;:debian/hurd\_test\_skip\_pipe \- http&#8203;://bugs\.debian\.org/650187
Disable failing GNU/Hurd tests io/pipe\.t
&nbsp;    DEBPKG&#8203;:debian/hurd\_test\_skip\_io\_pipe \-
http&#8203;://bugs\.debian\.org/650096 Disable failing GNU/Hurd tests
dist/IO/t/io\_pipe\.t
&nbsp;    DEBPKG&#8203;:fixes/CVE\-2012\-5195 \- avoid calling memset with a negative count
&nbsp;    DEBPKG&#8203;:fixes/CVE\-2012\-5526 \- \[PATCH 1/4\] CR escaping for P3P header
&nbsp;    DEBPKG&#8203;:CVE\-2013\-1667\.patch \- \[PATCH\] Prevent premature hsplit\(\)
calls\, and only trigger REHASH after hsplit\(\)
&nbsp;    DEBPKG&#8203;:CVE\-2012\-6329\.patch \-
http&#8203;://bugs\.debian\.org/cgi\-bin/bugreport\.cgi?bug=695224 \[1735f6f\] fix
arbitrary command execution via \_compile function in Maketext\.pm
&nbsp;  Built under linux
&nbsp;  Compiled at Feb  4 2014 23&#8203;:11&#8203;:19
&nbsp;  %ENV&#8203;:
&nbsp;    PERLBREW\_BASHRC\_VERSION="0\.67"
&nbsp;    PERLBREW\_HOME="/home/vse/\.perlbrew"
&nbsp;    PERLBREW\_MANPATH=""
&nbsp;    PERLBREW\_PATH="/home/perlbrew/bin"
&nbsp;    PERLBREW\_ROOT="/home/perlbrew"
&nbsp;    PERLBREW\_VERSION="0\.67"
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /etc/perl
&nbsp;    /usr/local/lib/perl/5\.14\.2
&nbsp;    /usr/local/share/perl/5\.14\.2
&nbsp;    /usr/lib/perl5
&nbsp;    /usr/share/perl5
&nbsp;    /usr/lib/perl/5\.14
&nbsp;    /usr/share/perl/5\.14
&nbsp;    /usr/local/lib/site\_perl
&nbsp;    \.

**Comment by p5pRT** [comment]
### From @druud62

On 2014\-05\-02 01&#8203;:23\, ÐÐ¸ÐºÑÐ¾Ñ ÐÑÐ¸Ð¼Ð¾Ð² wrote&#8203;:

> my $s1;
> $s1 = "x" x \(200\*1024\*1024\);
> print "ok\\n";
> sleep 1000;
> $s1 \.= '';
> ===
> Uses 400Mb \(during sleep\(\)\)
 >
 > reproduced in 5\.14 and 5\.19


This "similar thing" uses 200 MB in 5\.18\, but 100 MB in 5\.19&#8203;:

perl \-wE'
&nbsp;   open my $fh\, "\<"\, "100\_MB\.bin" or die $\!;
&nbsp;   my $s = do \{ local $/; \<$fh> \};
&nbsp;   sleep 1000;
' &



In 5\.18 it is better coded like&#8203;:

my $s; \{ local $/; $s = \<$fh> \}

\-\- 
Ruud

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @abigail

On Thu\, May 01\, 2014 at 04&#8203;:23&#8203;:33PM \-0700\, ÐÐ¸ÐºÑÐ¾Ñ ÐÑÐ¸Ð¼Ð¾Ð² wrote&#8203;:
> \# New Ticket Created by  ÐÐ¸ÐºÑÐ¾Ñ ÐÑÐ¸Ð¼Ð¾Ð² 
> \# Please include the string&#8203;:  \[perl \#121780\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=121780 >
> 
> 
> suboptimal memory usage of repetition operator
> 
> ===
> my $s1;
> $s1 \.= "x" for \(1\.\.200\*1024\*1024\);
> print "ok\\n";
> sleep 1000;
> $s1 \.= '';
> ===
> Uses 200Mb \(during sleep\(\)\)
> 
> 
> ===
> my $s1;
> $s1 = "x" x \(200\*1024\*1024\);
> print "ok\\n";
> sleep 1000;
> $s1 \.= '';
> ===
> Uses 400Mb \(during sleep\(\)\)
> 



That's because the latter contains "x" x \(200\*1024\*1024\) twice\.
It exists as a RHS value \(calculated at compile time\)\, which is then
copied into $s1\. Perl doesn't throw away the value\, as it doesn't
know whether it will be needed again\.

The former just has "x" on the RHS\, and that gets reused 200\*1024\*1024
times\.


You can reduce the memory usage of the latter by forcing Perl to release
the memory of the expression as soon as it's done by placing it in a
string eval\. But you'll pay if you execute it more than once in the life
time of the process&#8203;:

&nbsp;    my $s1;
&nbsp;    $s1 = eval '"x" x \(200\*1024\*1024\)';
&nbsp;    print "ok\\n";
&nbsp;    sleep 1000;
&nbsp;    $s1 \.= '';



Abigail

**Comment by p5pRT** [comment]
### From efimov@reg.ru

2014\-05\-02 12&#8203;:38 GMT\+04&#8203;:00 Abigail \<abigail@&#8203;abigail\.be>&#8203;:
> That's because the latter contains "x" x \(200\*1024\*1024\) twice\.
> It exists as a RHS value \(calculated at compile time\)\, which is then

at compile time? but same happening when it's runtime \(no constants\)&#8203;:

my $s1;
my $size = 200\*1024\*1024;
my $char = "x";
$s1 = $char x $size;
print "ok\\n";
sleep 1000;

**Comment by p5pRT** [comment]
### From @abigail

On Fri\, May 02\, 2014 at 12&#8203;:45&#8203;:05PM \+0400\, Victor Efimov wrote&#8203;:
> 2014\-05\-02 12&#8203;:38 GMT\+04&#8203;:00 Abigail \<abigail@&#8203;abigail\.be>&#8203;:
> > That's because the latter contains "x" x \(200\*1024\*1024\) twice\.
> > It exists as a RHS value \(calculated at compile time\)\, which is then
> 
> at compile time? but same happening when it's runtime \(no constants\)&#8203;:
> 
> my $s1;
> my $size = 200\*1024\*1024;
> my $char = "x";
> $s1 = $char x $size;
> print "ok\\n";
> sleep 1000;


Same thing happens\, except it now does it at run time\. It calculates
the RHS\, keeps the value\, and puts a copy of the value in $s1\.



Abigail

**Comment by p5pRT** [comment]
### From efimov@reg.ru

indeed\. That is \_not\_ issue when RHS side copied to LHS side and
immediately discarded \(but perl did not release memory back to OS\)\.

===
my $size = 200\*1024\*1024;
my $char = "x";
my $s1;
$s1 = $char x $size;
my $s2;
$s2 = $char x $size;
print "ok\\n";
===

\(800 Mb\)

Instead\, RHS of each statement saved for later re\-use \(through I don't
understand why it needed and how it can be reused\)\.


And that should take 600Mb then\, but it takes 800 too&#8203;:

===
my $size = 200\*1024\*1024;
my $char = "x";
\{
&nbsp;    my $s1;
&nbsp;    $s1 = $char x $size;
\}
\{
&nbsp;    my $s2;
&nbsp;    $s2 = $char x $size;
\}

print "ok\\n";
sleep 1000;

===



2014\-05\-02 12&#8203;:53 GMT\+04&#8203;:00 Abigail \<abigail@&#8203;abigail\.be>&#8203;:
> On Fri\, May 02\, 2014 at 12&#8203;:45&#8203;:05PM \+0400\, Victor Efimov wrote&#8203;:
>> 2014\-05\-02 12&#8203;:38 GMT\+04&#8203;:00 Abigail \<abigail@&#8203;abigail\.be>&#8203;:
>> > That's because the latter contains "x" x \(200\*1024\*1024\) twice\.
>> > It exists as a RHS value \(calculated at compile time\)\, which is then
>>
>> at compile time? but same happening when it's runtime \(no constants\)&#8203;:
>>
>> my $s1;
>> my $size = 200\*1024\*1024;
>> my $char = "x";
>> $s1 = $char x $size;
>> print "ok\\n";
>> sleep 1000;
>
>
> Same thing happens\, except it now does it at run time\. It calculates
> the RHS\, keeps the value\, and puts a copy of the value in $s1\.
>
>
>
> Abigail

**Comment by p5pRT** [comment]
### From @abigail

On Fri\, May 02\, 2014 at 01&#8203;:09&#8203;:18PM \+0400\, Victor Efimov wrote&#8203;:
> indeed\. That is \_not\_ issue when RHS side copied to LHS side and
> immediately discarded \(but perl did not release memory back to OS\)\.
> 
> ===
> my $size = 200\*1024\*1024;
> my $char = "x";
> my $s1;
> $s1 = $char x $size;
> my $s2;
> $s2 = $char x $size;
> print "ok\\n";
> ===
> 
> \(800 Mb\)
> 
> Instead\, RHS of each statement saved for later re\-use \(through I don't
> understand why it needed and how it can be reused\)\.


It keeps the internal structures around\, so it later doesn't have to
build them again\.

> 
> 
> And that should take 600Mb then\, but it takes 800 too&#8203;:
> 
> ===
> my $size = 200\*1024\*1024;
> my $char = "x";
> \{
>     my $s1;
>     $s1 = $char x $size;
> \}
> \{
>     my $s2;
>     $s2 = $char x $size;
> \}
> 
> print "ok\\n";
> sleep 1000;


Nothing is reused here\, you end up with two pairs of two copies&#8203;:
the first $char x $size\, and $s1\, and the second $char x $size and $s2\.


This\, however\, will take about 600Mb&#8203;:

&nbsp;    my $size = 200\*1024\*1024;
&nbsp;    my $char = "x";
&nbsp;    my @&#8203;s;
&nbsp;    for \(1 \.\. 2\) \{
&nbsp;        push @&#8203;s => $char x $size;
&nbsp;        print "ok\\n";
&nbsp;    \}
&nbsp;    sleep 1000;


Here we have three copies&#8203;: $char x $size\, $s \[0\] and $s \[1\]\.


Abigail

**Comment by p5pRT** [comment]
### From efimov@reg.ru

2014\-05\-02 13&#8203;:23 GMT\+04&#8203;:00 Abigail \<abigail@&#8203;abigail\.be>&#8203;:
> Nothing is reused here\, you end up with two pairs of two copies&#8203;:
> the first $char x $size\, and $s1\, and the second $char x $size and $s2\.

But isn't that a memory leak? $s1 goes out of scope and should be
discarded\, and its memory re\-used\.

**Comment by p5pRT** [comment]
### From @Leont

On Fri\, May 2\, 2014 at 10&#8203;:45 AM\, Victor Efimov \<efimov@&#8203;reg\.ru> wrote&#8203;:

> 2014\-05\-02 12&#8203;:38 GMT\+04&#8203;:00 Abigail \<abigail@&#8203;abigail\.be>&#8203;:
> > That's because the latter contains "x" x \(200\*1024\*1024\) twice\.
> > It exists as a RHS value \(calculated at compile time\)\, which is then
>
> at compile time? but same happening when it's runtime \(no constants\)&#8203;:
>
> my $s1;
> my $size = 200\*1024\*1024;
> my $char = "x";
> $s1 = $char x $size;
> print "ok\\n";
> sleep 1000;
>

You are correct that that is a bug\. It seems PADTMP's were not treated as
temporaries\, while TEMP sv's where \(as is evidenced by the eval in
Abigail's example\)\. Fortunately\, it seems Father C has already fixed this
in 9ffd39ab7\.

Leon

**Comment by p5pRT** [comment]
### From @iabyn

On Fri\, May 02\, 2014 at 10&#8203;:38&#8203;:16AM \+0200\, Abigail wrote&#8203;:
> That's because the latter contains "x" x \(200\*1024\*1024\) twice\.
> It exists as a RHS value \(calculated at compile time\)\,

I rather wonder whether we should not in fact be constant\-folding the
repeat operator\. For example

&nbsp;    if \($unlikely\_condition\) \{
&nbsp;        $x = 'x' x 200\_000\_000;
&nbsp;    \}
&nbsp;    else
&nbsp;        $x = '';
&nbsp;    \}

will consume 200Mb even if that branch is never taken\.

\-\- 
O Unicef Clearasil\!
Gibberish and Drivel\!
&nbsp;    \-\- "Bored of the Rings"

**Comment by p5pRT** [comment]
### From @abigail

On Fri\, May 02\, 2014 at 03&#8203;:22&#8203;:17PM \+0100\, Dave Mitchell wrote&#8203;:
> On Fri\, May 02\, 2014 at 10&#8203;:38&#8203;:16AM \+0200\, Abigail wrote&#8203;:
> > That's because the latter contains "x" x \(200\*1024\*1024\) twice\.
> > It exists as a RHS value \(calculated at compile time\)\,
> 
> I rather wonder whether we should not in fact be constant\-folding the
> repeat operator\. For example
> 
>     if \($unlikely\_condition\) \{
>         $x = 'x' x 200\_000\_000;
>     \}
>     else
>         $x = '';
>     \}
> 
> will consume 200Mb even if that branch is never taken\.


Yeah\, that's what I meant by "calculated at compile time"\.


I'm very aware of the effects of constant\-folding\, once bringing down
the website of $WORK\, because it ran out of memory due to code that
wasn't even executed \(but constant folded\)\.


Abigail

**Comment by p5pRT** [comment]
### From @lizmat

On 02 May 2014\, at 16&#8203;:27\, Abigail \<abigail@&#8203;abigail\.be> wrote&#8203;:
> On Fri\, May 02\, 2014 at 03&#8203;:22&#8203;:17PM \+0100\, Dave Mitchell wrote&#8203;:
>> On Fri\, May 02\, 2014 at 10&#8203;:38&#8203;:16AM \+0200\, Abigail wrote&#8203;:
>>> That's because the latter contains "x" x \(200\*1024\*1024\) twice\.
>>> It exists as a RHS value \(calculated at compile time\)\,
>> 
>> I rather wonder whether we should not in fact be constant\-folding the
>> repeat operator\. For example
>> 
>>    if \($unlikely\_condition\) \{
>>        $x = 'x' x 200\_000\_000;
>>    \}
>>    else
>>        $x = '';
>>    \}
>> 
>> will consume 200Mb even if that branch is never taken\.
> 
> Yeah\, that's what I meant by "calculated at compile time"\.
> 
> I'm very aware of the effects of constant\-folding\, once bringing down
> the website of $WORK\, because it ran out of memory due to code that
> wasn't even executed \(but constant folded\)\.

Yeah\.  Talking about side\-effects  :\-\)



Liz

**Comment by p5pRT** [comment]
### From @tsee

On 05/02/2014 04&#8203;:22 PM\, Dave Mitchell wrote&#8203;:
> On Fri\, May 02\, 2014 at 10&#8203;:38&#8203;:16AM \+0200\, Abigail wrote&#8203;:
>> That's because the latter contains "x" x \(200\*1024\*1024\) twice\.
>> It exists as a RHS value \(calculated at compile time\)\,
>
> I rather wonder whether we should not in fact be constant\-folding the
> repeat operator\. For example
>
>      if \($unlikely\_condition\) \{
>          $x = 'x' x 200\_000\_000;
>      \}
>      else
>          $x = '';
>      \}
>
> will consume 200Mb even if that branch is never taken\.

I think this is a corner case\. Few real programs have string constants 
that large and if they do\, they generally do for a reason\.

This is always a trade\-off\. Advocatus diaboli&#8203;: Should we stop allocation 
of scratchpads/targets because it's additional memory in a branch we 
never take? \(Obviously\, that would be a bad call\.\)

Constant folding is not rocket science\. People generally know it's going 
to happen and deal with the fallout\. The case that Abigail described was 
vastly more intricate than the above and it's the only such issue I've 
ever heard of\.

Maybe the proper solution \(in an ideal world\, I mean\) would be constant 
folding being performed at run\-time if the branch is first taken \(or 
only after it's taken N times\)\. Which gets kind of uncomfortably close 
to tracing JIT compilation\. \(I know about all the practical issues such 
as not allowing modification of OP trees at run time\.\)

\-\-Steffen

**Comment by p5pRT** [comment]
### From @avar

On Sat\, May 3\, 2014 at 11&#8203;:14 AM\, Steffen Mueller \<smueller@&#8203;cpan\.org> wrote&#8203;:

> The case that Abigail described was vastly more intricate
> than the above and it's the only such issue I've ever heard of\.

Are we remembering different issues? The one I remember came down to&#8203;:

&nbsp;    @&#8203;cache \{1 \.\. 999\_999\} = \(\);

I\.e\. just perl sizing the variable at compile\-time even though it
might not be used\. Hardly vastly more intricate\, or were you thinking
about some other case?

**Comment by p5pRT** [comment]
### From @abigail

On Sat\, May 03\, 2014 at 12&#8203;:44&#8203;:19PM \+0200\, Ãvar ArnfjÃ¶rÃ° Bjarmason wrote&#8203;:
> On Sat\, May 3\, 2014 at 11&#8203;:14 AM\, Steffen Mueller \<smueller@&#8203;cpan\.org> wrote&#8203;:
> 
> > The case that Abigail described was vastly more intricate
> > than the above and it's the only such issue I've ever heard of\.
> 
> Are we remembering different issues? The one I remember came down to&#8203;:
> 
>     @&#8203;cache \{1 \.\. 999\_999\} = \(\);
> 
> I\.e\. just perl sizing the variable at compile\-time even though it
> might not be used\. Hardly vastly more intricate\, or were you thinking
> about some other case?


That was the one\.



Abigail

**Comment by p5pRT** [comment]
### From @lizmat

On 03 May 2014\, at 21&#8203;:30\, Abigail \<abigail@&#8203;abigail\.be> wrote&#8203;:
> On Sat\, May 03\, 2014 at 12&#8203;:44&#8203;:19PM \+0200\, Ãvar ArnfjÃ¶rÃ° Bjarmason wrote&#8203;:
>> On Sat\, May 3\, 2014 at 11&#8203;:14 AM\, Steffen Mueller \<smueller@&#8203;cpan\.org> wrote&#8203;:
>> 
>>> The case that Abigail described was vastly more intricate
>>> than the above and it's the only such issue I've ever heard of\.
>> 
>> Are we remembering different issues? The one I remember came down to&#8203;:
>> 
>>    @&#8203;cache \{1 \.\. 999\_999\} = \(\);
>> 
>> I\.e\. just perl sizing the variable at compile\-time even though it
>> might not be used\. Hardly vastly more intricate\, or were you thinking
>> about some other case?
> 
> That was the one\.

Yup\, the one\.

Although the hash was \*not\* sized at compile time&#8203;: that happened at runtime using the array created by 1\.\.999999 at compile time\.  :\-\)

So the idea behind this was doing an optimisation at runtime\, with unfortunate severe side\-effects at compile time\.



Liz

**Comment by p5pRT** [comment]
### From @lizmat

On 03 May 2014\, at 21&#8203;:53\, Elizabeth Mattijsen \<liz@&#8203;dijkmat\.nl> wrote&#8203;:
> On 03 May 2014\, at 21&#8203;:30\, Abigail \<abigail@&#8203;abigail\.be> wrote&#8203;:
>> On Sat\, May 03\, 2014 at 12&#8203;:44&#8203;:19PM \+0200\, Ãvar ArnfjÃ¶rÃ° Bjarmason wrote&#8203;:
>>> On Sat\, May 3\, 2014 at 11&#8203;:14 AM\, Steffen Mueller \<smueller@&#8203;cpan\.org> wrote&#8203;:
>>> 
>>>> The case that Abigail described was vastly more intricate
>>>> than the above and it's the only such issue I've ever heard of\.
>>> 
>>> Are we remembering different issues? The one I remember came down to&#8203;:
>>> 
>>>   @&#8203;cache \{1 \.\. 999\_999\} = \(\);
>>> 
>>> I\.e\. just perl sizing the variable at compile\-time even though it
>>> might not be used\. Hardly vastly more intricate\, or were you thinking
>>> about some other case?
>> 
>> That was the one\.
> 
> Yup\, the one\.
> 
> Although the hash was \*not\* sized at compile time&#8203;: that happened at runtime using the array created by 1\.\.999999 at compile time\.  :\-\)
> 
> So the idea behind this was doing an optimisation at runtime\, with unfortunate severe side\-effects at compile time\.

Actually\, going down memory lane\, the very quick fix was&#8203;:

&nbsp;  eval â@&#8203;cache \{1 \.\. 999\_999\} = \(\)â;



Liz

**Comment by p5pRT** [comment]
### From @perhunter

On 05/03/2014 04&#8203;:14 PM\, Elizabeth Mattijsen wrote&#8203;:

>
> Actually\, going down memory lane\, the very quick fix was&#8203;:
>
>    eval â@&#8203;cache \{1 \.\. 999\_999\} = \(\)â;


why not a quick fix of&#8203;:

&nbsp;    $max = 999\_999 ;
&nbsp;    @&#8203;cache\{ 1 \.\. $max \} = \(\) ;

that shouldn't be able to trigger any compile time 'optimizations' of 
the generated list\.

uri

\-\- 
Uri Guttman \- The Perl Hunter
The Best Perl Jobs\, The Best Perl Hackers
http&#8203;://PerlHunter\.com

**Comment by p5pRT** [comment]
### From @tsee

On 05/03/2014 12&#8203;:44 PM\, Ãvar ArnfjÃ¶rÃ° Bjarmason wrote&#8203;:
> On Sat\, May 3\, 2014 at 11&#8203;:14 AM\, Steffen Mueller \<smueller@&#8203;cpan\.org> wrote&#8203;:
>
>> The case that Abigail described was vastly more intricate
>> than the above and it's the only such issue I've ever heard of\.
>
> Are we remembering different issues? The one I remember came down to&#8203;:
>
>      @&#8203;cache \{1 \.\. 999\_999\} = \(\);
>
> I\.e\. just perl sizing the variable at compile\-time even though it
> might not be used\. Hardly vastly more intricate\, or were you thinking
> about some other case?

Then I misremembered\. But again\, this is just a case of "don't do that" 
if you ask me\. And the above could rather easily be spotted\.

\-\-Steffen

**Comment by p5pRT** [comment]
### From @abigail

On Sun\, May 04\, 2014 at 09&#8203;:28&#8203;:48AM \+0200\, Steffen Mueller wrote&#8203;:
> On 05/03/2014 12&#8203;:44 PM\, Ãvar ArnfjÃ¶rÃ° Bjarmason wrote&#8203;:
>> On Sat\, May 3\, 2014 at 11&#8203;:14 AM\, Steffen Mueller \<smueller@&#8203;cpan\.org> wrote&#8203;:
>>
>>> The case that Abigail described was vastly more intricate
>>> than the above and it's the only such issue I've ever heard of\.
>>
>> Are we remembering different issues? The one I remember came down to&#8203;:
>>
>>      @&#8203;cache \{1 \.\. 999\_999\} = \(\);
>>
>> I\.e\. just perl sizing the variable at compile\-time even though it
>> might not be used\. Hardly vastly more intricate\, or were you thinking
>> about some other case?
>
> Then I misremembered\. But again\, this is just a case of "don't do that"  
> if you ask me\. And the above could rather easily be spotted\.
>


How many Perl programmers outside of p5p have any idea what perl is
doing at compile time? I'd wager most Perl programmers have no idea
what constant folding is\.

It's also rarely documented\. Perl\, the language\, is documented well\,
in man pages\, books and online\. But the runtime\, its optimizations\,
and its actions at compile time? Far less\.


Now\, I'm not arguing things should change\. But compile time actions
will keep biting people\. It certainly bit me\. 



Abigail

**Comment by p5pRT** [comment]
### From @bulk88

On Sat May 03 02&#8203;:14&#8203;:59 2014\, smueller@&#8203;cpan\.org wrote&#8203;:
> I think this is a corner case\. Few real programs have string constants 
> that large and if they do\, they generally do for a reason\.

I agree with that\. If someone will write something like that\, it has to have some legitimate purpose\. If you write something like that\, prepare for your process to have the available free memory to do that\. If you don't have the free memory to do that\, your accepting the risk of a random process termination and P5P can't help you with that\.

> This is always a trade\-off\. Advocatus diaboli&#8203;: Should we stop allocation 
> of scratchpads/targets because it's additional memory in a branch we 
> never take? \(Obviously\, that would be a bad call\.\)
> 
> Constant folding is not rocket science\. People generally know it's going 
> to happen and deal with the fallout\. The case that Abigail described was 
> vastly more intricate than the above and it's the only such issue I've 
> ever heard of\.
> 
> Maybe the proper solution \(in an ideal world\, I mean\) would be constant 
> folding being performed at run\-time if the branch is first taken \(or 
> only after it's taken N times\)\. Which gets kind of uncomfortably close 
> to tracing JIT compilation\. \(I know about all the practical issues such 
> as not allowing modification of OP trees at run time\.\)

I have a different idea but my recommendation on this ticket is to do nothing\, if the constant folding results in an SV with more than a 10 MB value\, reinstate the original optree\.

\-\- 
bulk88 ~ bulk88 at hotmail\.com
